### DISCUSSION TITLE
Possible checksum (ICV) error with Extended Sequence Numbers and replay_window 0

### DISCUSSION THREAD

**Discussion by IvanRancati** [strongswan/strongswan#2117] [Q&A] [answered]
I am currently debugging a problem with an appliance, based on Strongswan, and a VPN Gateway based on BSD
I have narrowed it down to this combination: esp_proposals contains ESN (Extended Sequence Number) and replay_window is 0
With this combination, an IPSec tunnel is established, but packets sent through the tunnel are not processed. 
I suspect this happens because the client sends an incorrect checksum (ICV), see below for my interpretation

So first the question: is this a correct scenario?

> esp_proposals = aes256-sha256-modp2048-esn\
> replay_window = 0

The following 3 combinations work without problems
          

> esp_proposals = aes256-sha256-modp2048-esn  

> esp_proposals = aes256-sha256-modp2048  

> esp_proposals = aes256-sha256-modp2048  
> replay_window = 0  

To have a testbed with only Strongswan and rule out BSD, I set up two Ubuntu VM following the "Road warrior" example
https://docs.strongswan.org/docs/5.9/config/quickstart.html#_roadwarrior_case
An extra ip 10.1.0.99 is added on moon ("server") to the ens3 interface, to have an IP to ping over the IPSec tunnel and send EPS packets

The configuration on moon (IP 192.168.1.221) accepts both ESN and NOESN:

>   connections {
>     rw {
>       proposals = aes256-sha256-modp2048
>       local {
>         auth = pubkey
>         certs = moonCert.pem
>         id = moon.strongswan.org
>       }
>       remote {
>         auth = pubkey
>       }
>       children {
>         rw {
>           local_ts  = 10.1.0.0/16
>           esp_proposals = aes256-sha256-modp2048-esn-noesn
>         }
>       }
>     }
>   }

Configuration on carol ("client", IP 192.168.1.116)
Here I have 4 different children combinations: with and without esn in esp_proposals, replay window not set (therefore with the default 32) or set to 0.

> conn-defaults {
>   remote_addrs = 192.168.1.221
>   proposals = aes256-sha256-modp2048
>   local {
>     auth = pubkey
>       certs = carolCert.pem
>       id = carol.strongswan.org
>     }
>   remote {
>     auth = pubkey
>     id = moon.strongswan.org
>   }
> }
> child-defaults {
>   remote_ts  = 10.1.0.0/16
>   start_action = none
> }
> 
> connections {
> 
>   esn-default : conn-defaults {
>     children {
>       esn-default : child-defaults {
>         esp_proposals = aes256-sha256-modp2048-esn
>       }
>     }
>   }
>   esn-no-window : conn-defaults {
>     children {
>       esn-no-window : child-defaults {
>         esp_proposals = aes256-sha256-modp2048-esn
>         replay_window = 0
>       }
>     }
>   }
>   no-esn : conn-defaults {
>     children {
>       no-esn : child-defaults {
>         esp_proposals = aes256-sha256-modp2048
>       }
>     }
>   }
>   no-esn-no-window : conn-defaults {
>     children {
>       no-esn-no-window : child-defaults {
>         esp_proposals = aes256-sha256-modp2048
>         replay_window = 0
>       }
>     }
>   }
> }

To test, from the client, with each of the 4 connections:
- if needed terminate other connections: 

> swanctl --terminate --ike  ...

- initiate the connection to be tested: 

> swanctl --initiate --child ...

- on the server, note the number of XfrmInStateProtoError packets 
    

>  grep Error /proc/net/xfrm_stat

- from the client, ping the extra address on the server
     

> ping -c 2 10.1.0.99 -s 37

With "esn-default", "no-esn", "no-esn-no-window", the server answers the ICMP packets. The tunnels are working
With "esn-no-window" the ICMP packets are not processed on the server. With     
  

> grep Error /proc/net/xfrm_stat

There are now 2 more XfrmInStateProtoError packets. No ICMP Answer

> root@konnektor-ms:/etc/swanctl# ping -c 2 10.1.0.99 -s 37
> PING 10.1.0.99 (10.1.0.99) 37(65) bytes of data.
> --- 10.1.0.99 ping statistics ---
> 2 packets transmitted, 0 received, 100% packet loss, time 1016ms

This matches the behaviour I saw using Strongswan as a VPN client and a BSD appliance as a VPN server: in that case, the BSD kernel rejected the packets because of incorrect checksums. 
Here the output of swanctl --initiate, ip xfrm state on client

> root@konnektor-ms:/etc/swanctl# swanctl --initiate --child esn-no-window
> [IKE] initiating IKE_SA esn-no-window[4] to 192.168.1.221
> [ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
> [NET] sending packet: from 192.168.1.116[500] to 192.168.1.221[500] (464 bytes)
> [NET] received packet: from 192.168.1.221[500] to 192.168.1.116[500] (497 bytes)
> [ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
> [CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
> [IKE] received cert request for "C=CH, O=strongSwan, CN=strongSwan Root CA"
> [IKE] sending cert request for "C=CH, O=strongSwan, CN=strongSwan Root CA"
> [IKE] authentication of 'carol.strongswan.org' (myself) with ED25519 successful
> [IKE] sending end entity cert "C=CH, O=strongswan, CN=carol.strongswan.org"
> [IKE] establishing CHILD_SA esn-no-window{5}
> [ENC] generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) N(HTTP_CERT_LOOK) CERTREQ IDr AUTH SA TSi TSr N(MOBIKE_SUP) N(NO_ADD_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
> [NET] sending packet: from 192.168.1.116[4500] to 192.168.1.221[4500] (768 bytes)
> [NET] received packet: from 192.168.1.221[4500] to 192.168.1.116[4500] (688 bytes)
> [ENC] parsed IKE_AUTH response 1 [ IDr CERT AUTH SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) ]
> [IKE] received end entity cert "C=CH, O=strongswan, CN=moon.strongswan.org"
> [CFG]   using certificate "C=CH, O=strongswan, CN=moon.strongswan.org"
> [CFG]   using trusted ca certificate "C=CH, O=strongSwan, CN=strongSwan Root CA"
> [CFG]   reached self-signed root ca with a path length of 0
> [CFG] checking certificate status of "C=CH, O=strongswan, CN=moon.strongswan.org"
> [CFG] certificate status is not available
> [IKE] authentication of 'moon.strongswan.org' with ED25519 successful
> [IKE] peer supports MOBIKE
> [IKE] IKE_SA esn-no-window[4] established between 192.168.1.116[carol.strongswan.org]...192.168.1.221[moon.strongswan.org]
> [IKE] scheduling rekeying in 13910s
> [IKE] maximum IKE_SA lifetime 15350s
> [CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/EXT_SEQ
> [IKE] CHILD_SA esn-no-window{5} established with SPIs cad9a83f_i c4469aa9_o and TS 192.168.1.116/32 === 10.1.0.0/16
> initiate completed successfully
> root@konnektor-ms:/etc/swanctl# ip xfrm state
> src 192.168.1.116 dst 192.168.1.221
>         proto esp spi 0xc4469aa9 reqid 1 mode tunnel
>         replay-window 0 flag af-unspec
>         auth-trunc hmac(sha256) 0x7f000afcb361d1dc27d096970122d4fc414e14d9ea7641e10e64bbd50d01b294 128
>         enc cbc(aes) 0x70bb26aafd89fe4855d921173ebdd9bc7aff1e3bd20d9c90471a5e8293bcfd06
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
> src 192.168.1.221 dst 192.168.1.116
>         proto esp spi 0xcad9a83f reqid 1 mode tunnel
>         replay-window 0 flag af-unspec
>         auth-trunc hmac(sha256) 0x3389e7cc833dbe62d1e33e1eb26e3ee656985cb400972d91a7548c1ebefbfa33 128
>         enc cbc(aes) 0x8e57fe1c45733576342abf6d143c0436872f5451e63d322fbb07e7de00942e91
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000

ip xfrm state on the server

> root@konnektor-krypto:~# ip xfrm state
> src 192.168.1.221 dst 192.168.1.116
>         proto esp spi 0xcad9a83f reqid 1 mode tunnel
>         replay-window 0 flag af-unspec esn
>         auth-trunc hmac(sha256) 0x3389e7cc833dbe62d1e33e1eb26e3ee656985cb400972d91a7548c1ebefbfa33 128
>         enc cbc(aes) 0x8e57fe1c45733576342abf6d143c0436872f5451e63d322fbb07e7de00942e91
>         anti-replay esn context:
>          seq-hi 0x0, seq 0x0, oseq-hi 0x0, oseq 0x0
>          replay_window 1, bitmap-length 1
>          00000000
> src 192.168.1.116 dst 192.168.1.221
>         proto esp spi 0xc4469aa9 reqid 1 mode tunnel
>         replay-window 0 flag af-unspec esn
>         auth-trunc hmac(sha256) 0x7f000afcb361d1dc27d096970122d4fc414e14d9ea7641e10e64bbd50d01b294 128
>         enc cbc(aes) 0x70bb26aafd89fe4855d921173ebdd9bc7aff1e3bd20d9c90471a5e8293bcfd06
>         anti-replay esn context:
>          seq-hi 0x0, seq 0x0, oseq-hi 0x0, oseq 0x0
>          replay_window 32, bitmap-length 1
>          00000000

Let me know if I can provide more details, or tcpdumps

As a side note: if you want to look at tcpdumps with Wireshark to evaluate the ICV, you will need a 4.3 build from https://www.wireshark.org/download/automated/, not the official 4.2.3 build, see issue https://gitlab.com/wireshark/wireshark/-/issues/19658

I am using strongswan 5.9.13 on Ubuntu 22.04.3 LTS, configured with the following options
--sysconfdir=/etc --enable-systemd --enable-swanctl --with-systemdsystemunitdir=/lib/systemd/system --disable-stroke

Thank you and best regards
Ivan Rancati

**Comment by tobiasbrunner** [comment]
> esp_proposals contains ESN (Extended Sequence Number) and replay_window is 0

This combination can't actually be configured in strongSwan on Linux. It will result in a non-ESN SA without replay protection. That's mainly because the Linux kernel currently doesn't support disabling replay protection with ESN.

**Reply by IvanRancati** [reply]
Thanks for the feedback. Basically, on Linux clients, this combination should never be used, right?

Just to make sure I understand correctly your comment. From the output of swanctl --initiate the connection seems to be ESN.
But the output of ip xfrm state shows no high bits for the Sequence Number. 
Would it be possible to log a warning /error on Linux clients, saying a connection has been initiated, but it's a non supported configuration? Or perhaps --initiate should fail.

Thanks and best regards

**Reply by tobiasbrunner** [reply, marked as answer]
> Basically, on Linux clients, this combination should never be used, right?

Yes, it currently can't be used. The kernel rejects SAs that have ESN enabled but don't configure a replay window.

> From the output of swanctl --initiate the connection seems to be ESN. But the output of ip xfrm state shows no high bits for the Sequence Number.

The CHILD_SA is negotiated with ESN. It's only a problem on the kernel level once it's installed. That's what's causing the issue as the two peers eventually don't agree on using ESN even if it was negotiated.

> Would it be possible to log a warning /error on Linux clients, saying a connection has been initiated, but it's a non supported configuration?

What's interesting is that until 8b9b11919d92e9738bb52901c9dbcc72e35b9fed there actually would have been an error during the installation of the SA. I actually think we should revert that commit because if ESN is negotiated it makes no sense to install the SA without it (unless strongSwan is running on both ends and the window was disabled on both).

**Reply by IvanRancati** [reply]
Just as a test, I modified line 2049 in src/libcharon/plugins/kernel_netlink/kernel_netlink_ipsec.c
 to 
> if (data->esn || data->replay_window > 32)

Now the connection with ESN and replay_window 0 fails (correctly so, in my opinion) to initiate. 

> [CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/EXT_SEQ
> [KNL] received netlink error: Invalid argument (22)
> [KNL] unable to add SAD entry with SPI c2bb6ad7 (FAILED)
> [KNL] received netlink error: Invalid argument (22)
> [KNL] unable to add SAD entry with SPI c0663c21 (FAILED)
> [IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel
> [IKE] failed to establish CHILD_SA, keeping IKE_SA
> [IKE] sending DELETE for ESP CHILD_SA with SPI c2bb6ad7
> [ENC] generating INFORMATIONAL request 2 [ D ]
> [NET] sending packet: from 192.168.1.116[4500] to 192.168.1.221[4500] (80 bytes)
> [NET] received packet: from 192.168.1.221[4500] to 192.168.1.116[4500] (80 bytes)
> [ENC] parsed INFORMATIONAL response 2 [ D ]
> [KNL] deleting policy 10.1.0.0/16 === 192.168.1.116/32 in failed, not found
> [KNL] deleting policy 10.1.0.0/16 === 192.168.1.116/32 fwd failed, not found
> initiate failed: establishing CHILD_SA 'esn-no-window' failed

Thanks again for all the information. From my point of view the problem is solved, but I'll be happy to run some more tests or try nightly builds if it helps.

**Reply by tobiasbrunner** [reply]
> Now the connection with ESN and replay_window 0 fails (correctly so, in my opinion) to initiate.

I agree. I've pushed a reversal of the mentioned commit to master (1f5aa8017f879ce6e5969196da62560586c121ef).
